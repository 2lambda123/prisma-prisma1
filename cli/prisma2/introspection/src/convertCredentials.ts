import { DatabaseCredentials } from './types'
import { DatabaseType } from 'prisma-datamodel'
import { URL } from 'url'
import { ConnectorType } from '@prisma/lift'

export function credentialsToUri(credentials: DatabaseCredentials): string {
  const type = databaseTypeToProtocol(credentials.type)
  if (credentials.type === DatabaseType.mongo) {
    return credentials.uri!
  }
  const url = new URL(type + '//')

  if (credentials.host) {
    url.hostname = credentials.host
  }

  if (credentials.type === DatabaseType.postgres) {
    if (credentials.database) {
      url.pathname = '/' + credentials.database
    }

    if (credentials.schema) {
      url.searchParams.set('schema', credentials.schema)
    }
  } else if (credentials.type === DatabaseType.mysql) {
    url.pathname = '/' + (credentials.database || credentials.schema)
  }

  if (credentials.user) {
    url.username = credentials.user
  }

  if (credentials.password) {
    url.password = credentials.password
  }

  if (credentials.port) {
    url.port = String(credentials.port)
  }

  return url.toString()
}

export function uriToCredentials(connectionString: string): DatabaseCredentials {
  const uri = new URL(connectionString)
  const type = protocolToDatabaseType(uri.protocol)

  // needed, as the URL implementation adds empty strings
  const exists = str => str && str.length > 0

  if (type === DatabaseType.mongo) {
    return {
      type,
      uri: connectionString, // todo: set authsource as database if not provided explicitly
    }
  }

  return {
    type,
    host: exists(uri.hostname) ? uri.hostname : undefined,
    user: exists(uri.username) ? uri.username : undefined,
    port: exists(uri.port) ? Number(uri.port) : undefined,
    password: exists(uri.password) ? uri.password : undefined,
    database: uri.pathname && uri.pathname.length > 1 ? uri.pathname.slice(1) : undefined,
    schema: uri.searchParams.get('schema') || undefined,
    uri: connectionString,
  }
}

function databaseTypeToProtocol(databaseType: DatabaseType): string {
  switch (databaseType) {
    case DatabaseType.postgres:
      return 'postgresql:'
    case DatabaseType.mysql:
      return 'mysql:'
    case DatabaseType.mongo:
      return 'mongodb:'
    case DatabaseType.sqlite:
      return 'sqlite:'
  }
}

function protocolToDatabaseType(protocol: string): DatabaseType {
  switch (protocol) {
    case 'postgresql:':
      return DatabaseType.postgres
    case 'mongodb:':
      return DatabaseType.mongo
    case 'mysql:':
      return DatabaseType.mysql
    case 'file:':
    case 'sqlite:':
      return DatabaseType.sqlite
  }

  throw new Error(`Unknown database type ${protocol}`)
}

export function databaseTypeToConnectorType(databaseType: DatabaseType): ConnectorType {
  switch (databaseType) {
    case DatabaseType.postgres:
      return 'postgres'
    case DatabaseType.mysql:
      return 'mysql'
    case DatabaseType.sqlite:
      return 'sqlite'
  }
  throw new Error(`Mongo is not yet supported`)
}
