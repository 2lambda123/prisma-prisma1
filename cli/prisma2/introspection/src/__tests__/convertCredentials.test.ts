import { uriToCredentials, credentialsToUri } from '../convertCredentials'

const uris = [
  'postgresql://',
  'postgresql://localhost',
  'postgresql://localhost:5433',
  'postgresql://localhost/mydb',
  'postgresql://user@localhost',
  'postgresql://user:secret@localhost',
  'postgresql://other@localhost/otherdb?schema=my_schema&connect_timeout=10&application_name=myapp',
  'mysql://user@localhost:3333',
  'mysql://user@localhost:3333/dbname',
  'mongodb://mongodb0.example.com:27017/admin',
  'mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin',
]

for (const uri of uris) {
  test(`Convert ${uri}`, () => {
    expect({ before: uri, after: credentialsToUri(uriToCredentials(uri)) }).toMatchSnapshot()
  })
}
