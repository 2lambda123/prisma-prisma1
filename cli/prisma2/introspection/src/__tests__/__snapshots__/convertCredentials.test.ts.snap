// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert mongodb://mongodb0.example.com:27017/admin 1`] = `
Object {
  "after": "mongodb://mongodb0.example.com:27017/admin",
  "before": "mongodb://mongodb0.example.com:27017/admin",
}
`;

exports[`Convert mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin 1`] = `
Object {
  "after": "mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin",
  "before": "mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin",
}
`;

exports[`Convert mysql://user@localhost:3333 1`] = `
Object {
  "after": "mysql://user@localhost:3333",
  "before": "mysql://user@localhost:3333",
}
`;

exports[`Convert mysql://user@localhost:3333/dbname 1`] = `
Object {
  "after": "mysql://user@localhost:3333/dbname",
  "before": "mysql://user@localhost:3333/dbname",
}
`;

exports[`Convert postgresql:// 1`] = `
Object {
  "after": "postgresql://",
  "before": "postgresql://",
}
`;

exports[`Convert postgresql://localhost 1`] = `
Object {
  "after": "postgresql://localhost",
  "before": "postgresql://localhost",
}
`;

exports[`Convert postgresql://localhost/mydb 1`] = `
Object {
  "after": "postgresql://localhost/mydb",
  "before": "postgresql://localhost/mydb",
}
`;

exports[`Convert postgresql://localhost:5433 1`] = `
Object {
  "after": "postgresql://localhost:5433",
  "before": "postgresql://localhost:5433",
}
`;

exports[`Convert postgresql://other@localhost/otherdb?schema=my_schema&connect_timeout=10&application_name=myapp 1`] = `
Object {
  "after": "postgresql://other@localhost/otherdb?schema=my_schema",
  "before": "postgresql://other@localhost/otherdb?schema=my_schema&connect_timeout=10&application_name=myapp",
}
`;

exports[`Convert postgresql://user:secret@localhost 1`] = `
Object {
  "after": "postgresql://user:secret@localhost",
  "before": "postgresql://user:secret@localhost",
}
`;

exports[`Convert postgresql://user@localhost 1`] = `
Object {
  "after": "postgresql://user@localhost",
  "before": "postgresql://user@localhost",
}
`;
